CDLUC3/dmsp_frontend_prototype
# Added some error handling ideas

This research was done as part of issue [#20].
(https://github.com/CDLUC3/dmsp_backend_prototype/issues/20)

I made some updates as examples of how we can handle errors:

- For graphQL requests:
-- Updated `lib/graphql/client.ts` with adding an `errorLink` to the client instance, and intercepting specific error types or codes to handle them differently if we want.
-- Updated `lib/graphql/server/queries/dmpDataQueries.ts` to 1) Map the error code to a user friendly message using a hash like `utils/userFriendlyErrorMessage`, log the error to the new `utils/logger.ts`, and throw an error that will be caught by the "try-catch" in the client- or server-side component. `logger.ts` can be used directly in client- and server-components as well.
-- Updated `dmps/[...slug]/page.tsx` to throw the returned error in the 'try-catch'
-- Installed `pino` and `@elastic/ecs-pino-format` to output formatted logs
- Added `app/not-found.tsx` to handle 404 errors 
- Added `app/error.tsx` to handle server-side errors. Client components will most likely display returned errors in the page using useState().


To test:

1. Start up app
2. navigate to `http://localhost:3001/dmps/10.48321/D1SP4H`
3. Look at the stack trace, and you will see the log: `{"log.level":"error","@timestamp":"2024-05-06T23:13:42.224Z","process.pid":4920,"host.hostname":"CDL-JSHIN-9M","ecs.version":"8.10.0","loggerPath":"getDMSP","message":"Error fetching item from DynamoDB"}`
